File: apps/ios/FishKit/Sources/PersonalChat/Logic/MessageMarkdown.swift
import Foundation

public enum MessageMarkdownInline: Sendable, Equatable {
    case text(String)
    case bold(String)
    case italic(String)
    case code(String)
    case link(label: String, href: String)
}

public struct MessageMarkdownListItem: Sendable, Equatable {
    public let content: [MessageMarkdownInline]
    public let children: MessageMarkdownList?

    public init(content: [MessageMarkdownInline], children: MessageMarkdownList? = nil) {
        self.content = content
        self.children = children
    }
}

public struct MessageMarkdownList: Sendable, Equatable {
    public let ordered: Bool
    public let items: [MessageMarkdownListItem]

    public init(ordered: Bool, items: [MessageMarkdownListItem]) {
        self.ordered = ordered
        self.items = items
    }
}

public enum MessageMarkdownBlock: Sendable, Equatable {
    case code(language: String?, content: String)
    case heading(level: Int, content: [MessageMarkdownInline])
    case blockquote(lines: [[MessageMarkdownInline]])
    case list(MessageMarkdownList)
    case paragraph(lines: [[MessageMarkdownInline]])
}

/// Deliberately small chat-markdown grammar ported from web. It never accepts
/// HTML and only marks http, https, and mailto destinations as links.
public enum MessageMarkdownParser {
    public static func parse(_ body: String) -> [MessageMarkdownBlock] {
        let lines = body.replacingOccurrences(of: "\r\n", with: "\n")
            .components(separatedBy: "\n")
        var blocks: [MessageMarkdownBlock] = []
        var index = 0

        while index < lines.count {
            let line = lines[index]
            if line.trimmingCharacters(in: .whitespaces).isEmpty {
                index += 1
                continue
            }

            if let language = fenceLanguage(line) {
                var code: [String] = []
                index += 1
                while index < lines.count, !isFenceClose(lines[index]) {
                    code.append(lines[index])
                    index += 1
                }
                if index < lines.count { index += 1 }
                blocks.append(.code(language: language.isEmpty ? nil : language, content: code.joined(separator: "\n")))
                continue
            }

            if let heading = heading(line) {
                blocks.append(.heading(level: heading.level, content: parseInline(heading.text)))
                index += 1
                continue
            }

            if quoteText(line) != nil {
                var quote: [[MessageMarkdownInline]] = []
                while index < lines.count, let text = quoteText(lines[index]) {
                    quote.append(parseInline(text))
                    index += 1
                }
                blocks.append(.blockquote(lines: quote))
                continue
            }

            if listItem(line) != nil {
                let parsed = parseList(lines, start: index, indent: 0)
                blocks.append(.list(parsed.list))
                index = parsed.nextIndex
                continue
            }

            var paragraph: [[MessageMarkdownInline]] = []
            while index < lines.count {
                let candidate = lines[index]
                if candidate.trimmingCharacters(in: .whitespaces).isEmpty || isBlockStart(candidate) {
                    break
                }
                paragraph.append(parseInline(candidate))
                index += 1
            }
            blocks.append(.paragraph(lines: paragraph))
        }
        return blocks
    }

    public static func sanitizedHref(_ value: String) -> String? {
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let separator = trimmed.firstIndex(of: ":") else { return nil }
        switch trimmed[..<separator].lowercased() {
        case "http", "https", "mailto": return trimmed
        default: return nil
        }
    }

    private static func parseList(
        _ lines: [String],
        start: Int,
        indent: Int
    ) -> (list: MessageMarkdownList, nextIndex: Int) {
        let ordered = listItem(lines[start]).map { $0.marker.last == "." } ?? false
        var items: [MessageMarkdownListItem] = []
        var index = start
        while index < lines.count {
            if lines[index].trimmingCharacters(in: .whitespaces).isEmpty {
                index += 1
                continue
            }
            guard let item = listItem(lines[index]) else { break }
            if item.indent < indent { break }
            if item.indent > indent {
                guard !items.isEmpty else { break }
                let nested = parseList(lines, start: index, indent: item.indent)
                let previous = items.removeLast()
                items.append(MessageMarkdownListItem(
                    content: previous.content,
                    children: nested.list
                ))
                index = nested.nextIndex
                continue
            }
            guard (item.marker.last == ".") == ordered else { break }
            items.append(MessageMarkdownListItem(content: parseInline(item.text)))
            index += 1
        }
        return (MessageMarkdownList(ordered: ordered, items: items), index)
    }

    private static func parseInline(_ text: String) -> [MessageMarkdownInline] {
        let pattern = #"`([^`]+)`|\*\*([^*]+)\*\*|\*([^*]+)\*|_([^_]+)_|\[([^\]]+)\]\(((?:[^()\s]|\([^()]*\))+)\)"#
        guard let regex = try? NSRegularExpression(pattern: pattern) else {
            return [.text(text)]
        }
        let fullRange = NSRange(text.startIndex..<text.endIndex, in: text)
        var result: [MessageMarkdownInline] = []
        var cursor = text.startIndex
        for match in regex.matches(in: text, range: fullRange) {
            guard let range = Range(match.range, in: text) else { continue }
            if cursor < range.lowerBound { result.append(.text(String(text[cursor..<range.lowerBound]))) }
            if let value = capture(1, from: match, in: text) {
                result.append(.code(value))
            } else if let value = capture(2, from: match, in: text) {
                result.append(.bold(value))
            } else if let value = capture(3, from: match, in: text) ?? capture(4, from: match, in: text) {
                result.append(.italic(value))
            } else if let label = capture(5, from: match, in: text),
                      let href = capture(6, from: match, in: text) {
                if let safe = sanitizedHref(href) {
                    result.append(.link(label: label, href: safe))
                } else {
                    result.append(.text(label))
                }
            }
            cursor = range.upperBound
        }
        if cursor < text.endIndex { result.append(.text(String(text[cursor...]))) }
        return result
    }

    private static func capture(
        _ group: Int,
        from match: NSTextCheckingResult,
        in text: String
    ) -> String? {
        guard match.range(at: group).location != NSNotFound,
              let range = Range(match.range(at: group), in: text) else { return nil }
        return String(text[range])
    }

    private static func fenceLanguage(_ line: String) -> String? {
        guard let match = firstMatch(#"^```\s*(\S*)\s*$"#, in: line) else { return nil }
        return capture(1, from: match, in: line) ?? ""
    }

    private static func isFenceClose(_ line: String) -> Bool {
        firstMatch(#"^```\s*$"#, in: line) != nil
    }

    private static func heading(_ line: String) -> (level: Int, text: String)? {
        guard let match = firstMatch(#"^(#{1,3})\s+(.*)$"#, in: line),
              let hashes = capture(1, from: match, in: line),
              let content = capture(2, from: match, in: line) else { return nil }
        return (hashes.count, content.trimmingCharacters(in: .whitespaces))
    }

    private static func quoteText(_ line: String) -> String? {
        guard let match = firstMatch(#"^>\s?(.*)$"#, in: line) else { return nil }
        return capture(1, from: match, in: line) ?? ""
    }

    private static func listItem(_ line: String) -> (indent: Int, marker: String, text: String)? {
        guard let match = firstMatch(#"^(\s*)([-*]|\d+\.)\s+(.*)$"#, in: line),
              let spacing = capture(1, from: match, in: line),
              let marker = capture(2, from: match, in: line),
              let content = capture(3, from: match, in: line) else { return nil }
        return (spacing.count, marker, content)
    }

    private static func isBlockStart(_ line: String) -> Bool {
        fenceLanguage(line) != nil || heading(line) != nil || quoteText(line) != nil || listItem(line) != nil
    }

    private static func firstMatch(_ pattern: String, in text: String) -> NSTextCheckingResult? {
        guard let regex = try? NSRegularExpression(pattern: pattern) else { return nil }
        return regex.firstMatch(
            in: text,
            range: NSRange(text.startIndex..<text.endIndex, in: text)
        )
    }
}

File: apps/android/feature/chat/src/main/kotlin/space/fishhub/android/feature/chat/logic/MessageMarkdown.kt
package space.fishhub.android.feature.chat.logic

sealed interface MessageMarkdownInline {
    data class Text(val text: String) : MessageMarkdownInline
    data class Bold(val text: String) : MessageMarkdownInline
    data class Italic(val text: String) : MessageMarkdownInline
    data class Code(val text: String) : MessageMarkdownInline
    data class Link(val label: String, val href: String) : MessageMarkdownInline
}

data class MessageMarkdownListItem(
    val content: List<MessageMarkdownInline>,
    val children: MessageMarkdownList? = null,
)

data class MessageMarkdownList(val ordered: Boolean, val items: List<MessageMarkdownListItem>)

sealed interface MessageMarkdownBlock {
    data class Code(val language: String?, val content: String) : MessageMarkdownBlock
    data class Heading(val level: Int, val content: List<MessageMarkdownInline>) : MessageMarkdownBlock
    data class Blockquote(val lines: List<List<MessageMarkdownInline>>) : MessageMarkdownBlock
    data class Bullets(val list: MessageMarkdownList) : MessageMarkdownBlock
    data class Paragraph(val lines: List<List<MessageMarkdownInline>>) : MessageMarkdownBlock
}

/**
 * Deliberately small chat-markdown grammar ported from web
 * (`message-body-parser.ts`) and iOS (`MessageMarkdown.swift`) so all three
 * clients render the same subset. It never accepts HTML and only marks http,
 * https, and mailto destinations as links.
 */
object MessageMarkdownParser {

    private val FenceOpenRe = Regex("""^```\s*(\S*)\s*$""")
    private val FenceCloseRe = Regex("""^```\s*$""")
    private val HeadingRe = Regex("""^(#{1,3})\s+(.*)$""")
    private val BlockquoteRe = Regex("""^>\s?(.*)$""")
    private val ListItemRe = Regex("""^(\s*)([-*]|\d+\.)\s+(.*)$""")
    private val InlineRe = Regex(
        """`([^`]+)`|\*\*([^*]+)\*\*|\*([^*]+)\*|_([^_]+)_|\[([^\]]+)\]\(((?:[^()\s]|\([^()]*\))+)\)""",
    )

    fun parse(body: String): List<MessageMarkdownBlock> {
        val lines = body.replace("\r\n", "\n").split("\n")
        val blocks = mutableListOf<MessageMarkdownBlock>()
        var index = 0

        while (index < lines.size) {
            val line = lines[index]
            if (line.trim().isEmpty()) {
                index += 1
                continue
            }

            val language = fenceLanguage(line)
            if (language != null) {
                val code = mutableListOf<String>()
                index += 1
                while (index < lines.size && !isFenceClose(lines[index])) {
                    code.add(lines[index])
                    index += 1
                }
                if (index < lines.size) index += 1
                blocks.add(
                    MessageMarkdownBlock.Code(
                        language = if (language.isEmpty()) null else language,
                        content = code.joinToString("\n"),
                    ),
                )
                continue
            }

            val headingMatch = heading(line)
            if (headingMatch != null) {
                blocks.add(MessageMarkdownBlock.Heading(headingMatch.level, parseInline(headingMatch.text)))
                index += 1
                continue
            }

            if (quoteText(line) != null) {
                val quote = mutableListOf<List<MessageMarkdownInline>>()
                while (index < lines.size) {
                    val text = quoteText(lines[index]) ?: break
                    quote.add(parseInline(text))
                    index += 1
                }
                blocks.add(MessageMarkdownBlock.Blockquote(quote))
                continue
            }

            if (listItem(line) != null) {
                val parsed = parseList(lines, index, 0)
                blocks.add(MessageMarkdownBlock.Bullets(parsed.list))
                index = parsed.nextIndex
                continue
            }

            val paragraph = mutableListOf<List<MessageMarkdownInline>>()
            while (index < lines.size) {
                val candidate = lines[index]
                if (candidate.trim().isEmpty() || isBlockStart(candidate)) break
                paragraph.add(parseInline(candidate))
                index += 1
            }
            blocks.add(MessageMarkdownBlock.Paragraph(paragraph))
        }
        return blocks
    }

    fun sanitizedHref(value: String): String? {
        val trimmed = value.trim()
        val scheme = trimmed.substringBefore(':', missingDelimiterValue = "")
        return when (scheme.lowercase()) {
            "http", "https", "mailto" -> trimmed
            else -> null
        }
    }

    /** Flattens the parsed tree to plain text, for an accessibility string. */
    fun plainText(body: String): String = parse(body).joinToString("\n\n") { blockPlainText(it) }

    private fun blockPlainText(block: MessageMarkdownBlock): String = when (block) {
        is MessageMarkdownBlock.Code -> block.content
        is MessageMarkdownBlock.Heading -> inlinePlainText(block.content)
        is MessageMarkdownBlock.Blockquote -> block.lines.joinToString("\n") { inlinePlainText(it) }
        is MessageMarkdownBlock.Bullets -> listPlainText(block.list)
        is MessageMarkdownBlock.Paragraph -> block.lines.joinToString("\n") { inlinePlainText(it) }
    }

    private fun listPlainText(list: MessageMarkdownList): String =
        list.items.joinToString("\n") { item ->
            val own = inlinePlainText(item.content)
            val children = item.children?.let(::listPlainText)
            if (children.isNullOrEmpty()) own else "$own\n$children"
        }

    private fun inlinePlainText(inlines: List<MessageMarkdownInline>): String =
        inlines.joinToString("") { inline ->
            when (inline) {
                is MessageMarkdownInline.Text -> inline.text
                is MessageMarkdownInline.Bold -> inline.text
                is MessageMarkdownInline.Italic -> inline.text
                is MessageMarkdownInline.Code -> inline.text
                is MessageMarkdownInline.Link -> inline.label
            }
        }

    private data class ListParseResult(val list: MessageMarkdownList, val nextIndex: Int)

    private fun parseList(lines: List<String>, start: Int, indent: Int): ListParseResult {
        val ordered = listItem(lines[start])?.marker?.endsWith(".") ?: false
        val items = mutableListOf<MessageMarkdownListItem>()
        var index = start
        while (index < lines.size) {
            if (lines[index].trim().isEmpty()) {
                index += 1
                continue
            }
            val item = listItem(lines[index]) ?: break
            if (item.indent < indent) break
            if (item.indent > indent) {
                if (items.isEmpty()) break
                val nested = parseList(lines, index, item.indent)
                val previous = items.removeAt(items.size - 1)
                items.add(previous.copy(children = nested.list))
                index = nested.nextIndex
                continue
            }
            if (item.marker.endsWith(".") != ordered) break
            items.add(MessageMarkdownListItem(content = parseInline(item.text)))
            index += 1
        }
        return ListParseResult(MessageMarkdownList(ordered, items), index)
    }

    private fun parseInline(text: String): List<MessageMarkdownInline> {
        val result = mutableListOf<MessageMarkdownInline>()
        var cursor = 0
        for (match in InlineRe.findAll(text)) {
            val range = match.range
            if (cursor < range.first) {
                result.add(MessageMarkdownInline.Text(text.substring(cursor, range.first)))
            }
            val code = capture(match, 1)
            val bold = capture(match, 2)
            val italic = capture(match, 3) ?: capture(match, 4)
            val label = capture(match, 5)
            val href = capture(match, 6)
            when {
                code != null -> result.add(MessageMarkdownInline.Code(code))
                bold != null -> result.add(MessageMarkdownInline.Bold(bold))
                italic != null -> result.add(MessageMarkdownInline.Italic(italic))
                label != null && href != null -> {
                    val safe = sanitizedHref(href)
                    result.add(
                        if (safe != null) {
                            MessageMarkdownInline.Link(label, safe)
                        } else {
                            MessageMarkdownInline.Text(label)
                        },
                    )
                }
            }
            cursor = range.last + 1
        }
        if (cursor < text.length) {
            result.add(MessageMarkdownInline.Text(text.substring(cursor)))
        }
        return result
    }

    private fun capture(match: MatchResult, group: Int): String? = match.groups[group]?.value

    private fun fenceLanguage(line: String): String? = FenceOpenRe.find(line)?.groupValues?.get(1)

    private fun isFenceClose(line: String): Boolean = FenceCloseRe.matches(line)

    private data class HeadingMatch(val level: Int, val text: String)

    private fun heading(line: String): HeadingMatch? {
        val match = HeadingRe.find(line) ?: return null
        return HeadingMatch(match.groupValues[1].length, match.groupValues[2].trim())
    }

    private fun quoteText(line: String): String? = BlockquoteRe.find(line)?.groupValues?.get(1)

    private data class ListItemMatch(val indent: Int, val marker: String, val text: String)

    private fun listItem(line: String): ListItemMatch? {
        val match = ListItemRe.find(line) ?: return null
        return ListItemMatch(match.groupValues[1].length, match.groupValues[2], match.groupValues[3])
    }

    private fun isBlockStart(line: String): Boolean =
        fenceLanguage(line) != null || heading(line) != null || quoteText(line) != null || listItem(line) != null
}
